gate-oauth2/src/main/java/com/netflix/spinnaker/gate/security/oauth2/SpinnakerUserInfoTokenServices.java

```diff
@@ -0,0 +1,304 @@
+
+  private RetrySupport retrySupport = new RetrySupport();
+
+  @Override
+  public OAuth2Authentication loadAuthentication(final String accessToken)
+      throws AuthenticationException, InvalidTokenException {
+    OAuth2Authentication oAuth2Authentication =
+        userInfoTokenServices.loadAuthentication(accessToken);
+
+    final Map<String, Object> details =
+        (Map<String, Object>) oAuth2Authentication.getUserAuthentication().getDetails();
+
+    log.debug("UserInfo details: " + details);
+
+    boolean isServiceAccount = isServiceAccount(details);
+    if (!isServiceAccount) {
+      if (!hasAllUserInfoRequirements(details)) {
+        throw new BadCredentialsException("User's info does not have all required fields.");
+      }
+
+      if (providerTokenServices != null
+          && !providerTokenServices.hasAllProviderRequirements(accessToken, details)) {
+        throw new BadCredentialsException(
+            "User's provider info does not have all required fields.");
+      }
+    }
+
+    final String username = safeGet(details.get(userInfoMapping.getUsername()));
+    List<String> roles =
+        Optional.ofNullable(groupExtractor)
+            .map(extractor -> extractor.apply(accessToken, details))
+            .orElseGet(
+                () -> Optional.ofNullable(getRoles(details)).orElse(Collections.emptyList()));
+    // Service accounts are already logged in.
+    if (!isServiceAccount) {
+      var id = registry.createId("fiat.login").withTag("type", "oauth2");
+
+      try {
+        retrySupport.retry(
+            () -> {
+              if (roles.isEmpty()) {
+                permissionService.login(username);
+              } else {
+                permissionService.loginWithRoles(username, roles);
+              }
+              return "Successfully Logged in" + username;
+            },
+            5,
+            Duration.ofSeconds(2),
+            false);
+        log.debug(
+            "Successful oauth2 authentication (user: {}, roleCount: {}, roles: {})",
+            username,
+            roles.size(),
+            roles);
+        id = id.withTag("success", true).withTag("fallback", "none");
+      } catch (Exception e) {
+        log.debug(
+            "Unsuccessful oauth2 authentication (user: {}, roleCount: {}, roles: {}, legacyFallback: {})",
+            username,
+            roles.size(),
+            roles,
+            fiatClientConfigurationProperties.isLegacyFallback(),
+            e);
+        id =
+            id.withTag("success", false)
+                .withTag("fallback", fiatClientConfigurationProperties.isLegacyFallback());
+
+        if (!fiatClientConfigurationProperties.isLegacyFallback()) {
+          throw e;
+        }
+      } finally {
+        registry.counter(id).increment();
+      }
+    }
+
+    User spinnakerUser = new User();
+    spinnakerUser.setEmail(safeGet(details.get(userInfoMapping.getEmail())));
```

Comment:
same question about a possible behavior change here with safeGet.

Reply:
I renamed the safeGet method to toStringOrNull and implemented groovy equivalent functionality as part of the same method.

Reply:
https://github.com/spinnaker/gate/pull/1871/commits/51348ba555a2cc53a433626517d4c706d0fa4e49